### ISSUE TITLE
excessive memory usage in more

### ISSUE THREAD

**Issue by mjguzik** [uutils/coreutils#6397]
Running it over a test file of about 1G in size results in 15G RSS.

Problems start with slurping the entire thing upfront, I don't know how it manages to amass 15 x memory overhead on top of it.

more from gnu coreutils sits at around 2MB.

How to repro:
1. generate a file of about 1G: `perl -e 'print "meh\n" x (256 * 1024 * 1024)' > testfile`
2. `cargo run -r testfile`

**Comment by sylvestre** [comment]
note that we have been focusing on compatibility, not perf or memory usage for now (even if some programs are already doing better than GNU's)

**Comment by tertsdiepraam** [comment]
True, but I also agree that 15x the filesize is a bit excessive 😄

**Comment by pstef** [comment]
Even 100% would be excessive, consumed memory shouldn't scale linearly with the file size at all.

**Comment by tertsdiepraam** [comment]
It's also just extremely slow to load such a file. I'm looking into making this at least a bit better.

**Comment by mjguzik** [comment]
> note that we have been focusing on compatibility, not perf or memory usage for now (even if some programs are already doing better than GNU's)

For coreutils sensible memory usage is part of correctness, especially so for a tool like more.

For example say I logged in to a production machine with the intent of running `more /var/log/log_file_sized_1G_or_more`. Suppose the the explosive memory usage problem is fixed, but the entire file is still slurped upfront, resulting in 1G RSS.
If the machine I'm running this on does not have a free 1G we are running into trouble. And chances are things are already going wrong, hence why I'm logging in to poke around (including reading logs) in the first place.

The gnu coreutils variant uses few MB of ram making itself a perfectly viable tool in that scenario.

That is to say this code needs to deal with the file in chunks and at the moment is not fit for use whatsoever. Any work done on it which would be thrown away while making the adjustment is a waste of time, which would be better spent either reworking this or beating other tools to shape in a lasting manner.

**Comment by tertsdiepraam** [comment]
Not useful for you does not mean useless for everyone. We've already agreed this is an issue and want to fix it. Your point has been made.

**Comment by mmstick** [comment]
It is most likely the case that high memory usage in coreutils—and most other Rust applications—is simply because most don't know that GNU libc malloc requires manual trim calls to release unused memory from its brk arenas. There is also  by default a dynamic mmap threshold that quickly escalates the size of brk allocations as Rust allocates memory through the system allocator. Either you should set a reasonable static mmap threshold, or manually allocate large buffers with mmap directly. See this for example, as used by libcosmic 
 https://github.com/pop-os/libcosmic/blob/master/src/malloc.rs

**Comment by tertsdiepraam** [comment]
While that might be a fix elsewhere, it's definitely not the main issue here. The issue is mostly that `more` reads in the entire file and splits it into a `Vec` of lines. That's unnecessary and also a very inefficient representation (1 time file size + a pointer and length for each line). So there's more pressing memory concerns than fiddling with the memory allocation. It should just store only the  part of the file currently on screen.

I'm curious whether you find other utils where this might help though. Feel free to open a more general issue about it for awareness.

By the way, if somebody sees this, it's a really fun first issue to optimize this!

**Comment by aaron-ang** [comment]
hi @tertsdiepraam, I'm interested in working on this

**Comment by tertsdiepraam** [comment]
@aaron-ang Great! Go for it!

**Comment by aaron-ang** [comment]
After the PR above is merged, I will improve `more`'s initialization memory overhead. Right now it still scales linearly with file size. We will work on making it linear :)

**Comment by aaron-ang** [comment]
Is there a manual for the GNU `more` implementation? The [man page](https://man7.org/linux/man-pages/man1/more.1.html) shows very different options than the ones in the code.

**Comment by tertsdiepraam** [comment]
`more` shouldn't be part of coreutils, it's actually from util-linux, I think. Maybe we'll set that up sometime, but for now it lives here. Remember not to look at GNU source code for licensing reasons by the way.

**Comment by aaron-ang** [comment]
@tertsdiepraam looking at the `more` manual again, I realized it does not support backward line scrolling (only page). This feature is implemented in `less`. We have sort of a hybrid implementation, which makes refactoring the code tricky. So to clarify, are we looking to match our implementation with `more` or `less`? or remain as a hybrid?
